src/topology/homotopy/path.lean

```diff
@@ -0,0 +1,148 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shing Tak Lam
+-/
+
+import topology.homotopy.basic
+import topology.path_connected
+
+/-!
+# Homotopy between paths
+
+In this file, we define a `homotopy` between two `path`s. In addition, we define a relation
+`homotopic` on `path`s`, and prove that it is an equivalence relation. Finally, we also define
+(but do not yet prove that is a group) the fundamental group of a space.
+
+## Definitions
+
+* `path.homotopy p₀ p₁` is the type of homotopies between paths `p₀` and `p₁`
+* `path.homotopy.refl p` is the constant homotopy between `p` and itself
+* `path.homotopy.symm F` is the `path.homotopy p₁ p₀` defined by reversing the homotopy
+* `path.homotopy.trans F G`, where `F : path.homotopy p₀ p₁`, `G : path.homotopy p₁ p₂` is the
+  `path.homotopy p₀ p₂` defined by putting the first homotopy on `[0, 1/2]` and the second on
+  `[1/2, 1]`
+* `path.homotopic p₀ p₁` is the relation saying that there is a homotopy between `p₀` and `p₁`
+* `path.homotopic.setoid x₀ x₁` is the setoid on `path`s from `path.homotopic`
+* `path.homotopic.quotient x₀ x₁` is the quotient type from `path x₀ x₀` by `path.homotopic.setoid`
+* `first_homotopy x₀` is defined to be `path.homotopic.quotient x₀ x₀`. That is, the fundamental
+  group of `X` based at `x₀`.
+
+## Todos
+
+Prove that `first_homotopy x₀` is a `group`.
+-/
+
+universes u v
+
+variables {X : Type u} {Y : Type v} [topological_space X] [topological_space Y]
+variables {x₀ x₁ : X}
+
+noncomputable theory
+
+open_locale unit_interval
+
+namespace path
+
+/--
+The type of homotopies between two paths.
+-/
+structure homotopy (p₀ p₁ : path x₀ x₁) extends
+  continuous_map.homotopy p₀.to_continuous_map p₁.to_continuous_map :=
+(source' : ∀ t, to_fun (t, 0) = x₀)
+(target' : ∀ t, to_fun (t, 1) = x₁)
+
+namespace homotopy
+
+section
+
+variables {p₀ p₁ : path x₀ x₁}
+
+instance : has_coe_to_fun (homotopy p₀ p₁) := ⟨_, λ p, p.to_fun⟩
+
+@[simp]
+lemma to_homotopy_apply (h : homotopy p₀ p₁) {t : I × I} :
+  h.to_homotopy t = h t := rfl
+
+@[simp]
+lemma source (F : homotopy p₀ p₁) {t : I} : F (t, 0) = x₀ := F.source' t
+
+@[simp]
+lemma target (F : homotopy p₀ p₁) {t : I} : F (t, 1) = x₁ := F.target' t
+
+@[ext]
+lemma ext {F G : homotopy p₀ p₁} (h : ∀ t, F t = G t) : F = G :=
```

Comment:
Can you prove `@function.injective (homotopy p₀ p₁) sorry coe_fn` too, where the sorry is the type that the coercion has (I can't work it out from the source alone).

Reply:
I'll give that a go. Should I prove it for `continuous_map.homotopy` as well then?

Reply:
It's shorter to prove `ext` in terms of `coe_fn_injective` (and as a bonus avoids quot.sound in `coe_fn_injective`)

Reply:
> (and as a bonus avoids quot.sound in coe_fn_injective)

am I missing something? since if I just have

```lean
lemma coe_fn_injective : @function.injective (homotopy f₀ f₁) (I × X → Y) coe_fn := sorry

#print axioms coe_fn_injective
```

this has `quot.sound` already. But I'll change the order around so that `ext` uses `coe_fn_injective`.